[quant][graphmode] Support ops with fixed quantization parameters

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38436 [quant] Histogram Observer support List[Tensor]
* #38435 [quant] Remove get_qparams in Observers
* #38434 [quant][graphmode] Add RemoveReduantDequantize pass
* #38359 [quant][graphmode][refactor] Factor out getFixedQParamOpFusionInfo
* #38358 [quant][graphmode][refactor] Factor out getInputTensorQParamOpFusionInfo
* **#38278 [quant][graphmode] Support ops with fixed quantization parameters**

Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21559811](https://our.internmc.facebook.com/intern/diff/D21559811)

jerryzh168 committed:
[quant][graphmode] Support ops with fixed quantization parameters

Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -133,6 +139,31 @@ std::vector<std::string> _single_input_general_value_aten_funcs = {
     "leaky_relu_",
 };
 
+std::tuple<c10::QScheme, QParamVector> _per_tensor_asym_qparam =
```
can we add some comments on where the numerical values are coming from?  Even if we can't reuse the constants directly, some context for future code readers on where they are coming from would be helpful

vkuzo commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -133,6 +139,31 @@ std::vector<std::string> _single_input_general_value_aten_funcs = {
+std::tuple<c10::QScheme, QParamVector> _per_tensor_asym_qparam =
+    std::make_tuple(
+        c10::kPerTensorAffine,
+        QParamVector({std::make_pair(".scale", IValue(1.0f / 256.0f)),
```
+1 to lint

vkuzo commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -688,6 +688,94 @@ graph(%a_quant, %negative_slope):
 
+  // IR pattern common to all ops with fixed quantization parameters for
+  // asymetric quantization
+  std::string asym_fixed_qparam_op_suffix = R"(
```
can we also add some comments on the magic numbers here, even if we can't pull them out into constants?

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21559811](https://our.internmc.facebook.com/intern/diff/D21559811)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support ops with fixed quantization parameters"


Summary:
Support ops like aten::hardsigmoid that has a fixed quantization parameters:
```
  constexpr float o_scale = 1.0f / 256.0f;
  constexpr int32_t o_zero_point = 0;
```

Ops supported:
- hardsigmoid
- sigmoid
- tanh

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21559811](https://our.internmc.facebook.com/intern/diff/D21559811)

[ghstack-poisoned]
